doc/unboxing.md

```diff
@@ -0,0 +1,234 @@
+Unboxing
+========
+
+## Overview
+
+The unboxing algorithm in Flambda 2 aims to remove unnecessary allocations
+without causing bad behaviour (e.g. pushing allocations into loops).
+It currently supports:
+
+- Unboxing along non-looping control flow, within a single function, of the
+  following varieties of values:
+  - Boxed `float`
+  - Boxed `int32`
+  - Boxed `int64`
+  - Boxed `nativeint`
+  - Variants
+  - Closures (need examples).
+
+- Unboxing around loops, save that the corresponding allocations will only be
+  removed if:
+  - the first use of the value being unboxed in the loop can make use of the
+    unboxed form (for example if it is a floating-point arithmetic
+    operation, as opposed to something like a non-inlined function call); and
+  - this same condition also holds for the first use of the value after
+    the loop, on all exit paths from the loop.  (The trick of "`+. 0.`" can
+    be used here; this will not be optimised away by Flambda 2.)
+
+The unboxing code does not handle the untagging of integers.
+
+In the future support is expected for:
+
+- Unboxed returns from functions (e.g. returning values of variant type
+  without allocating).
+
+- Removal of allocations from loops where there does not always exist a use
+  of the corresponding unboxed value subsequent to the loop.  This
+  optimisation is expected to be based on partial dead code elimination.
+
+## Control over unboxing
+
+Unboxing depth.
```

Comment:
There are other control knobs to turn on/off parts of the unboxing process (cf https://github.com/ocaml-flambda/ocaml/pull/394/files#diff-b476bdab198a007fdfdcefdcd15ef528f6a86274223c89bab6a909d2be93187fR33-R36 ), although they are not exposed and should really only be used in debugging scenarios.